drivers/net/wireless/marvell/mwifiex/cfg80211.c

```diff
@@ -962,6 +962,138 @@ mwifiex_init_new_priv_params(struct mwifiex_private *priv,
+			return adapter->curr_iface_comb.p2p_intf ==
+			       adapter->iface_limit.p2p_intf;
+		case NL80211_IFTYPE_AP:
+			return adapter->curr_iface_comb.uap_intf ==
+			       adapter->iface_limit.uap_intf;
+		default:
+			return false;
+		}
+
+	case NL80211_IFTYPE_STATION:
+		switch (new_iftype) {
+		case NL80211_IFTYPE_ADHOC:
+			return true;
+		case NL80211_IFTYPE_P2P_CLIENT:
+		case NL80211_IFTYPE_P2P_GO:
+			return adapter->curr_iface_comb.p2p_intf ==
+			       adapter->iface_limit.p2p_intf;
+		case NL80211_IFTYPE_AP:
+			return adapter->curr_iface_comb.uap_intf ==
+			       adapter->iface_limit.uap_intf;
+		default:
+			return false;
+		}
+
+	case NL80211_IFTYPE_AP:
+		switch (new_iftype) {
+		case NL80211_IFTYPE_ADHOC:
+			return adapter->curr_iface_comb.sta_intf ==
+			       adapter->iface_limit.sta_intf;
+		case NL80211_IFTYPE_P2P_CLIENT:
+		case NL80211_IFTYPE_P2P_GO:
+			return adapter->curr_iface_comb.p2p_intf ==
+			       adapter->iface_limit.p2p_intf;
+		default:
+			return false;
+		}
+
+	case NL80211_IFTYPE_P2P_CLIENT:
+		switch (new_iftype) {
+		case NL80211_IFTYPE_ADHOC:
+		case NL80211_IFTYPE_STATION:
+			return true;
+		case NL80211_IFTYPE_P2P_GO:
+			return true;
+		case NL80211_IFTYPE_AP:
+			return adapter->curr_iface_comb.uap_intf ==
+			       adapter->iface_limit.uap_intf;
+		default:
+			return false;
+		}
+
+	case NL80211_IFTYPE_P2P_GO:
+		switch (new_iftype) {
+		case NL80211_IFTYPE_ADHOC:
+		case NL80211_IFTYPE_STATION:
+			return true;
+		case NL80211_IFTYPE_P2P_CLIENT:
+			return true;
+		case NL80211_IFTYPE_AP:
+			return adapter->curr_iface_comb.uap_intf ==
+			       adapter->iface_limit.uap_intf;
+		default:
+			return false;
+		}
+
+	default:
+		break;
+	}
+
+	return false;
+}
+
+static void
+update_vif_type_counters(struct mwifiex_adapter *adapter,
+			 enum nl80211_iftype old_iftype,
+			 enum nl80211_iftype new_iftype)
+{
+	if (old_iftype != -1) {
+		switch (old_iftype) {
+		case NL80211_IFTYPE_UNSPECIFIED:
+		case NL80211_IFTYPE_ADHOC:
+		case NL80211_IFTYPE_STATION:
+			adapter->curr_iface_comb.sta_intf--;
+			break;
+		case NL80211_IFTYPE_AP:
+			adapter->curr_iface_comb.uap_intf--;
+			break;
+		case NL80211_IFTYPE_P2P_CLIENT:
+		case NL80211_IFTYPE_P2P_GO:
+			adapter->curr_iface_comb.p2p_intf--;
+			break;
+		default:
+			/* This should be dead code; checked above */
+			mwifiex_dbg(adapter, ERROR, "type not supported\n");
+			break;
+		}
+	}
+
+	if (new_iftype != -1) {
+		switch (new_iftype) {
+		case NL80211_IFTYPE_UNSPECIFIED:
+		case NL80211_IFTYPE_ADHOC:
+		case NL80211_IFTYPE_STATION:
+			adapter->curr_iface_comb.sta_intf++;
+			break;
+		case NL80211_IFTYPE_AP:
+			adapter->curr_iface_comb.uap_intf++;
+			break;
+		case NL80211_IFTYPE_P2P_CLIENT:
+		case NL80211_IFTYPE_P2P_GO:
+			adapter->curr_iface_comb.p2p_intf++;
+			break;
+		default:
+			/* This should be dead code; checked above */
+			mwifiex_dbg(adapter, ERROR, "type not supported\n");
+			break;
+		}
+	}
+}
```

Comment:
I think it would be cleaner to split this into two functions, one for incrementing and one for decrementing (removing the need for the special `-1` value). Also you can split out a third function that updates the counters based on an input value, i.e. `update(type, value)` and then use `update(type, +1)` and `update(type, -1)` inside some small wrappers for inc and dec (basically just for naming).

Reply:
Went with only one helper function that gets passed `+1`/`-1` directly now, adding increase/decrease wrappers seemed a little too much IMHO.